Source/Particles/Collision/BinaryCollision/VirtualPhotonCreation.H

```diff
@@ -0,0 +1,248 @@
+
+#include <AMReX_INT.H>
+#include <AMReX_REAL.H>
+#include <AMReX_Particle.H>
+
+#include <cmath>
+
+
+namespace collision::binarycollision::virtualphotons{
+
+using namespace amrex::literals;
+using SoaData_type = typename WarpXParticleContainer::ParticleTileType::ParticleTileDataType;
+
+
+/** Generate virtual photons from primary particles (electrons or positrons)
+ * using the equivalent photon approximation, also known as the Weizsacker-Williams method.
+ * See for example Jackson, Classical Electrodynamics, 3rd edition, section 15.4
+ * or Berestetskii, Lifshitz, Pitaevskii, Quantum Electrodynamics, 2nd edition, section 99.
+ *
+ * The number of photons per primary particle is computed according to Berestetskii's formula 99.16,
+ * integrated over angles and over photon energies from a minimum energy (input parameter)
+ * up to the energy of the primary particle.
+ * The photons are assumed to be emitted in the direction of motion of the primary particle.
+ * The virtual photons are generated in a separate species, which can then be used
+ * as input for other processes, e.g., Breit-Wheeler pair production and Compton scattering.
+ * The virtual photons are generated at the same position as the primary particle,
+ * and inherit the same weight as the primary particle, divided by a sampling factor.
+ * The virtual photons are given the same particle ID as the primary particle.
+ * The virtual photon species is regenerated from scratch at each time step.
+ * In the current implementation, the primary particles are not affected by
+ * the possible interactions that the virtual photons may undergo.
+ *
+ * \param[in] mypc MultiParticleContainer calling this method
+ */
+void GenerateVirtualPhotons (MultiParticleContainer* mypc){
+
+    WARPX_PROFILE("collision::binarycollision::virtualphotons::GenerateVirtualPhotons()");
+
+    // Loop through the species
+    for (int i_s = 0; i_s < mypc->nSpecies(); ++i_s) {
+
+        auto& primary = mypc->GetParticleContainer(i_s);
+
+        if(!primary.has_virtual_photons()){
+            continue;
+        }
+
+        // Get the virtual photon species corresponding to this primary species
+        int vphotons_index = primary.getVirtualPhotonSpeciesIndex();
+        auto& vphotons = mypc->GetParticleContainer(vphotons_index);
+        const amrex::ParmParse pp_species_name(mypc->GetSpeciesNames()[vphotons_index]);
+
+        // Minimum allowed enrgy of the virtual photons
+        amrex::Real vphoton_min_energy;
+        utils::parser::getWithParser(pp_species_name, "virtual_photons_min_energy", vphoton_min_energy);
+
+        // Sampling factor:
+        // the number of virtual photons generated is multiplied by this factor,
+        // the weight of each virtual photon is divided by this factor
+        amrex::Real sampling_factor;
+        utils::parser::getWithParser(pp_species_name, "sampling_factor", sampling_factor);
+
+        // Enable tiling
+        amrex::MFItInfo info;
```

Comment:
Could this be the reason it fails when `particles.do_tiling = 0`? It looks like this loop always has tiling enabled. But those tiles might not exist in the underyling particle containers. I think we only want to enable tiling in this loop if `particles.do_tiling = 1`.